src/eva/core/callbacks/writers/embeddings.py

```diff
@@ -122,38 +125,105 @@ def _get_embeddings(self, prediction: torch.Tensor) -> torch.Tensor:
 
-    def _construct_save_name(self, input_name, metadata, local_idx):
-        group_name = metadata[self._group_key][local_idx] if self._group_key else None
-        save_name = os.path.splitext(input_name)[0] + ".pt"
-        if group_name:
-            save_name = os.path.join(group_name, save_name)
-        return input_name, save_name
+    def _get_item_metadata(
+        self, metadata: Dict[str, Any] | None, local_idx: int
+    ) -> Dict[str, Any] | None:
+        """Returns the metadata for the item at the given local index."""
+        if not metadata:
+            if self._metadata_keys:
+                raise ValueError("Metadata keys are provided but the batch metadata is empty.")
+            else:
+                return None
+
+        item_metadata = {}
+        for key in self._metadata_keys:
+            if key not in metadata:
+                raise KeyError(f"Metadata key '{key}' not found in the batch metadata.")
+            item_metadata[key] = metadata[key][local_idx]
+
+        return item_metadata
 
 
 def _process_write_queue(
-    write_queue: multiprocessing.Queue, output_dir: str, overwrite: bool = False
+    write_queue: multiprocessing.Queue,
+    output_dir: str,
+    metadata_keys: List[str],
+    save_every_n: int,
+    overwrite: bool = False,
 ) -> None:
-    manifest_file, manifest_writer = _init_manifest(output_dir, overwrite)
+    manifest_file, manifest_writer = _init_manifest(output_dir, metadata_keys, overwrite)
+
+    save_name_to_items: Dict[str, ITEM_DICT_ENTRY] = {}
+
+    counter = 0
     while True:
         item = write_queue.get()
         if item is None:
             break
 
-        prediction_buffer, target_buffer, input_name, save_name, split, slide_id = QUEUE_ITEM(*item)
-        _save_prediction(prediction_buffer, save_name, output_dir)
-        _update_manifest(target_buffer, input_name, save_name, split, slide_id, manifest_writer)
+        item = QUEUE_ITEM(*item)
+
+        if item.save_name in save_name_to_items:
+            save_name_to_items[item.save_name].items.append(item)
+        else:
+            save_name_to_items[item.save_name] = ITEM_DICT_ENTRY(items=[item], save_count=0)
+
+        if counter > 0 and counter % save_every_n == 0:
+            save_name_to_items = _save_items(
+                save_name_to_items, metadata_keys, output_dir, manifest_writer
+            )
+
+        counter += 1
+
+    if len(save_name_to_items) > 0:
+        _save_items(save_name_to_items, metadata_keys, output_dir, manifest_writer)
 
     manifest_file.close()
 
 
-def _save_prediction(prediction_buffer: io.BytesIO, save_name: str, output_dir: str) -> None:
-    save_path = os.path.join(output_dir, save_name)
-    prediction = torch.load(io.BytesIO(prediction_buffer.getbuffer()), map_location="cpu")
+def _save_items(
+    save_name_to_items: Dict[str, ITEM_DICT_ENTRY],
+    metadata_keys: List[str],
+    output_dir: str,
+    manifest_writer: Any,
+) -> Dict[str, ITEM_DICT_ENTRY]:
```

Comment:
even though this is wip, i would already add a docstring with description. The logic (together with `save_predictions`) is quite tricky and a concise description might make it easier to follow.

Reply:
good point, I now added a couple of docstrings to explain what happens.